x/bank/module.go

```diff
@@ -3,50 +3,68 @@ package bank
 import (
 	"encoding/json"
 
+	"github.com/gorilla/mux"
+	"github.com/spf13/cobra"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
 	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
 	"github.com/cosmos/cosmos-sdk/x/auth"
-	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/cosmos/cosmos-sdk/x/bank/client/rest"
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
 )
 
 var (
-	_ sdk.AppModule      = AppModule{}
-	_ sdk.AppModuleBasic = AppModuleBasic{}
+	_ module.AppModule      = AppModule{}
+	_ module.AppModuleBasic = AppModuleBasic{}
 )
 
-// name of this module
-const ModuleName = "bank"
-
 // app module basics object
 type AppModuleBasic struct{}
 
-var _ sdk.AppModuleBasic = AppModuleBasic{}
+var _ module.AppModuleBasic = AppModuleBasic{}
 
 // module name
 func (AppModuleBasic) Name() string {
-	return ModuleName
+	return types.ModuleName
 }
 
 // register module codec
 func (AppModuleBasic) RegisterCodec(cdc *codec.Codec) {
-	RegisterCodec(cdc)
+	types.RegisterCodec(cdc)
 }
 
 // default genesis state
 func (AppModuleBasic) DefaultGenesis() json.RawMessage {
-	return moduleCdc.MustMarshalJSON(DefaultGenesisState())
+	return types.ModuleCdc.MustMarshalJSON(DefaultGenesisState())
 }
 
 // module validate genesis
 func (AppModuleBasic) ValidateGenesis(bz json.RawMessage) error {
 	var data GenesisState
-	err := moduleCdc.UnmarshalJSON(bz, &data)
+	err := types.ModuleCdc.UnmarshalJSON(bz, &data)
 	if err != nil {
 		return err
 	}
 	return ValidateGenesis(data)
 }
 
+// register rest routes
+func (AppModuleBasic) RegisterRESTRoutes(ctx context.CLIContext, rtr *mux.Router, cdc *codec.Codec) {
+	rest.RegisterRoutes(ctx, rtr, cdc)
+}
+
+// TODO
```

Comment:
What is actually missing here?

Reply:
client and query commands are manually added for bank in gaia, we do this because "sending" is considered root level functionality within the `gaiacli` it would be good to rethink how commands are ordered in the CLI and be able to perform this ordering in the module manager level. But that's for the future, for now I'm going to delete these `TODO`s

Reply:
How about opening an issue instead?

Reply:
I'm not sure this even is an issue tbh, like maybe the existing design is fine. I'll open an issue if it seems like there is a clear problem to address, but I'm not sure the existing structure is even problematic, it may already be the optimum structure, just allow for manual deviation from the module client pattern for special outlier commands like `send` instead of building this customization into the module manager. leaving for now

Reply:
I'm in favor of moving this to `bank`. An SDK app w/ a UTXO base would not be able to use any of this code. Also that command relies entirely on bank.

Reply:
ref https://github.com/cosmos/cosmos-sdk/pull/4451#discussion_r290901624